### PR TITLE
fix(core): Fix user comparison in same-user subworkflow caller policy

### PR DESCRIPTION
**Pull request by ivov**
https://linear.app/n8n/issue/PAY-992

https://community.n8n.io/t/executing-workflow-using-owner-role-created-by-another-user-fails/33443

### CODE REVIEW

**Review by krynble — APPROVED** [review]
LGTM!
Thanks for the fix!

**Review by krynble — CHANGES_REQUESTED** [review]
We still have a problem with this PR - the error thrown by `checkSubworkflowExecutePolicy` is of type `SubworkflowOperationError`.

The problem is that the caller function expects the error to be of type `ExecutionError`.

Long story short: the `SubworkflowOperationError` error type is missing a `node` attribute stating the failed node, so we can show the error correctly in the proper node.

**Comment by ivov** [comment]
> We still have a problem with this PR - the error thrown by `checkSubworkflowExecutePolicy` is of type `SubworkflowOperationError`.
> 
> The problem is that the caller function expects the error to be of type `ExecutionError`.
> 
> Long story short: the `SubworkflowOperationError` error type is missing a `node` attribute stating the failed node, so we can show the error correctly in the proper node.

Thanks! Quick question: Is this in scope of this PR? Looks like a preexisting issue in multiple places whenever `SubworkflowOperatorError` is used. Might be best to ship the fix and address this separately, else let me know.

**Comment by krynble** [comment]
@ivov we'll have to fix this as the scope of this PR otherwise the error message is still completely useless as it errors with `Failed reading property of undefined: reading name` or something similar.

On the other hand this should not be too hard to fix, as it's a matter of extending the correct error class and sending the`Execute Workflow` node used to trigger. Hopefully this won't be too complicated.

**Comment by ivov** [comment]
Thanks! I'll come back to this soon.

File: packages/cli/src/UserManagement/PermissionChecker.ts [hunk 1]
```diff
@@ -101,15 +99,40 @@ export class PermissionChecker {
...
+
+		try {
+			parentWorkflow = await Container.get(WorkflowRepository).findOneOrFail({
+				where: { id: parentWorkflowId },
+			});
+		} catch (error) {
+			throw new ApplicationError('Failed to find parent workflow', { extra: { parentWorkflowId } });
+		}
+
+		const executeWorkflowNode = Object.values(parentWorkflow.nodes).find(
```
Comment (krynble) [inline packages/cli/src/UserManagement/PermissionChecker.ts:119]: The problem I see here is that if a workflow contains more than 1 execute workflow node, we'll get the first one we find.

Instead, what we could do is pass the `node` from https://github.com/n8n-io/n8n/blob/f5502cc628f6b348f7fe3325b96ec9dc3360beaf/packages/core/src/NodeExecuteFunctions.ts#L3155 to this function call. Then we'd have the exact node that was trying to execute a child workflow and display the error in the proper place.

File: packages/cli/src/WorkflowHelpers.ts [hunk 2]
```diff
@@ -175,8 +175,7 @@ export async function executeErrorWorkflow(
 		try {
 			await PermissionChecker.checkSubworkflowExecutePolicy(
 				workflowInstance,
-				runningUser.id,
-				workflowErrorData.workflow.id,
+				workflowErrorData.workflow.id!,
```
Comment (krynble) [inline packages/cli/src/WorkflowHelpers.ts:178]: In here we could use `workflowInstance.getNode(workflowErrorData.execution?.lastNodeExecuted);` if available as it should contain the last executed node.

**Review by krynble — APPROVED** [review]
I've made a small change to `WorkflowHelpers.ts`

**Comment by janober** [comment]
Got released with [`n8n@1.21.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.21.0)

**Pull request merged**